<div class='main'>
  <h2>Diabetes Risk</h2>
  <br/>
  <div>
    <button id='compute-diabetes' type='submit'>
      Compute my Type 2 Diabetes Risk
    </button>
    <div id='please-choose-another-population' title='Please Select Another Population' class='required hidden'>
      This exercise is not implemented for population: <span id='pop-error-box'></span>. Please choose another population.
    </div>
  </div>
  
  <br />
  <br />
  
  <div id='diabetes-chart'></div>
  
  <br />
  
  <table id='diabetes-table' class='results-table hidden'>
    <thead>
    <tr>
      <th>dbSNP</th>
      <th>Genotype</th>
      <th>Study size</th>
      <th>Imputed from</th>
      <th>R squared</th>
      <th>LR</th>
      <th>Running LR</th>
      <th>Probability</th>
    </tr>
    </thead>
    <tbody>
    <script id='diabetes-snp-template' type='text/template'>
      <tr>
        <td class='key'><%= dbsnp %></td>
        <td class='value'><%= genotype %></td>
        <td><%= study_size %></td>
        <td><%= imputed_from %></td>
        <td><%= r_squared %></td>
      </tr>
    </script>
    </tbody>
  </table>
  
  <div class='hidden'>
    <br />
    <button id='submit-diabetes'>
      Submit my information
    </button>
  </div>
  
  <dialog id='confirm-submit-diabetes' title='Submit my information' class='dialog hidden'>
    Submit your information?
  </dialog>
  
  <dialog id='submitted-diabetes' title='Thanks' class='dialog hidden'>
    <strong>Submitted!</strong>
  </dialog>
  
</div>

<div class='help'>
  <div id='help-diabetes-help'>
    <h2>Clinical Assessment</h2>
    Type 2 diabetes mellitus is a complex polygenic trait and is the most common form of diabetes.
    Heritability is estimated at about 20%, but thus far, variants discovered by GWAS have only been able to explain a fraction of the heritability.
    <br/><br/>
    In this exercise, you can profile your risk for this common disease using a few of the SNPs identified by association studies.
    We will use the method demonstrated by the <a href='http://www.ncbi.nlm.nih.gov/pubmed/20435227'>first clinical assessment of a personal genome</a>.
    For this assessment, we require likelihood ratios conferred by each SNP (which are not always provided by the initial GWAS).
    The likelihood ratio is then multiplied by your prior probability (given the background population selected) to obtain a posterior likelihood.
    This likelihood represents the likelihood of the disease given the data observed so far, and is converted to a probability, which is shown here.
    This process is repeated for all the SNPs in the dataset (ordered by discovery study size) to obtain your final probability.
    <br/><br/>
    Note that this does not include all the SNPs associated with the disease (nor are all these SNPs even known, as the heritability is not yet fully explained).
    Note that because of this:
    <br/><br/>
    <li>
      This analysis is <strong>incomplete</strong>, as plenty more variants may be identified
    </li>
    <li>
      This analysis <strong>may differ</strong> from other interpretations offered by other companies (such as 23andme) or programs, which may consider different SNPs for their interpretation.
    </li>
    <br/>
    Again, note that this information is not intended as a diagnosis.
    First, even a full clinical genomic assessment would not constitute a diagnosis, as these variants simply alter the disease risk of an individual.
    In addition, this exercise does not consider all variants, but simply illustrates current methods for genome interpretation.
    As always, consult your physician for medical advice.
    <br/><br/>
  </div>
</div>

<br class='clear' />