<div class='main'>
  <h2>Lookup</h2>
  <br/>
  <div>
    <h4>Enter SNPs to lookup (one per line)</h4>
    <textarea id='lookup-snps-textarea' rows='5' cols='30'></textarea>
    <br/>
    <br/>
    <button id='lookup-snps' type='submit' class='small-button'>
      Lookup SNPs
    </button>
  </div>
  <br class='clear' />
  <div id='table-options'>
    <br/>
    <!--<button id='submit-snps' type='submit'>-->
    <!--  Submit my SNPs-->
    <!--</button>-->
    <button id='clear-snps' type='submit' class='small-button'>
      Clear Table
    </button>
    <button id='toggle-unknown-genotypes' type='submit' class='small-button'>
      Hide unknown genotypes
    </button>
  </div>
	
  <!-- Submission buttons. -->
  <div id='too-many-snps' title='Too Many' class='dialog hidden'>
    You have entered too many SNPs. Your browser will not be able to display this many SNPs.
  </div>
  
	<br />
	
	<!-- General lookup template. -->
	<table id='lookup-snps-table' class='results-table hidden'>
    <thead>
	  <tr>
	    <th class='dbsnp'>dbSNP</th><th class='genotype'>Genotype</th>
      <th class='reference details'>Reference</th><th class='refallele_freq details'>Reference<br/>Frequency</th>
      <th class='other details'>Alternate<br/>Allele</th><th class='otherallele_freq details'>Alternate<br/>Frequency</th>
      <th class='chrom details'>Chrom.</th><th class='position details'>Position</th>
      <th class='imputed'>Imputed From</th><th class='rsq'>R<sup>2</sup></th>
	  </tr>
    </thead>
    <tbody>
	  <script id='lookup-snp-template' type='text/template'>
		  <tr>
		    <td class='dbsnp key'><%= dbsnp %></td>
        <td class='genotype value'><%= genotype %></td>
        <td class='reference details'><%= reference %></td>
        <td class='refallele_freq details'><%= refallele_freq %></td>
        <td class='other details'><%= otherallele %></td>
        <td class='other_freq details'><%= otherallele_freq %></td>
        <td class='chrom details'><%= chrom %></td>
        <td class='position details'><%= end %></td>
        <td class='imputed'><%= imputed_from %> </td>
        <td class='rsq'><%= r_squared %></td>
		  </tr>
		</script>
    </tbody>
	</table>
	
  <br/>
  
  <!-- Lookup demo. -->
  <h2 class='link secondary-color' id='lookup-demo'>Imputation Demo</h2>
  <div class='hidden'>
    <h3>Enter SNPs to delete (one per line)</h3>
    Delete some SNPs that are already in your genotype file.
    <div>
      <textarea id='delete-snps-textarea' rows='3' cols='15'></textarea>
      <br/>
      <button id='delete-snps' type='submit' class='small-button'>
        Remove SNPs from Genome
      </button>
    </div>
    Then, try to lookup the same SNPs and observe whether they were imputed correctly.
    <br/>
  </div>
  
  <br/>
  
  <!-- BED file. -->
  <h2 class='link secondary-color' id='lookup-bed'>Output in Bed Format</h2>
  <div class='hidden'>
    <li>
      Open the <a href="http://genome.ucsc.edu/cgi-bin/hgGateway">UCSC Genome Browser</a> to visualize your data.
    </li>
    <li>
      In the "assembly" dropdown, choose "Mar. 2006 (NCBI36/hg18)."
    </li>
    <li>
      Click "add custom tracks" and paste the contents of this box into the "Paste URLs or Data" box and click "Submit."
    </li>
    <li>
      Note that by doing so, you are uploading genotype data to UCSC.
      These data are anonymous, but these genotypes are no longer kept secure in your browser.
    </li>
    <li>
      Click "Go to Genome Browser" and explore. The top track ('Personal Genotype') will contain these SNPs.
    </li>
    <li>
      They are labeled as red for homozygous alternate, blue for heterozygous, and black for homozygous reference.
    </li>
    <script id='bed-file-template' type='text/template'>chr<%= chrom %> <%= start %> <%= end %> rs<%= dbsnp %>(<%= genotype %>) 0 + <%= start %> <%= end %> <%= color %>
</script>
    <textarea id='bed-file-text' rows=5 cols=70>track name=myGenome description="Personal Genotype" visibility=3 itemRgb="On"
</textarea>
  </div>
</div>

<div class='help'>
  <div id='help-lookup-help'>
    <h2>Lookup</h2>
    This utility looks up SNPs in your personal genotype file.
    <br><br>
    If the SNP is not found in your file, the utility attempts to impute the SNP using Hapmap data.
    This requires:
    <li>
      The SNP you request to be found in the Hapmap population requested
    </li>
    <li>
      Another SNP (the "linked" SNP) in high LD (r<sup>2</sup> &#x2265 cutoff) with the requested SNP.
    </li>
    <li>
      This linked SNP to be found both in Hapmap and your own data
    </li>
    <br>
    <h3>Process</h3>
    First, we find a SNP that is in high linkage with the requested SNP. This SNP must be found both in Hapmap and your data.
    <br/><br/>
    The Hapmap project has determined the "phase" of these SNPs, which means a correspondence can be determined between alleles.
    Thus, the presence of one allele at the first SNP indicates the presence of the allele with the same phase of the other SNP.
    <br/><br/>
    Click the "explain" button on an imputed SNP to see an example.
    If "??" is displayed, the SNP could not be imputed with the data given and population selected.
    <br/>
    <br/>
  </div>
  <div class='description'>
    <div>
      <div id='explain-lookup-top'>
        <script id='explain-lookup-top-template' type='text/template'>
          &nbsp;&nbsp;&nbsp;
          <%= dbsnp %>
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
          <%= imp_dbsnp %>
          <br/>
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
          <%= genotype_0 %>
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
          &nbsp;&nbsp;&nbsp;&nbsp;
          <%= imp_0 %>
        </script>
      </div>
        ------|-------------------------|-------
        <br/>
        ------|-------------------------|-------
      <div id='explain-lookup-bottom'>
        <script id='explain-lookup-bottom-template' type='text/template'>
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
          <%= genotype_1 %>
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
          &nbsp;&nbsp;&nbsp;&nbsp;
          <%= imp_1 %>
        </script>
      </div>
    </div>
  </div>
</div>
<br class='clear' />
